ripple/src/main/webapp/trie.js

```diff
@@ -0,0 +1,103 @@
+/* @source https://gist.github.com/tpae/72e1c54471e88b689f85ad2b3940a8f0 */
+
+/* Create TrieNode class. Used in Trie implementation */
+var Trie = function () {
+  var that = Object.create(Trie.prototype);
+  that.children = {}; // mapping: next character -> child nodes
+  that.isWord = false;
+  
+  /* Insert one word. */
+  that.insertWord = function (word) {
+    var current_node = that;
+    for (var i = 0; i < word.length; i++) {
+      var c = word[i]
+      // if character is not in the trie already, add it
+      if (!(c in current_node.children)) {
+        current_node.children[c] = Trie();
+      }
+      // update current_node
+      current_node = current_node.children[c];
+    };
+
+      // after adding all the chars of the word,
+      // you are at the end of a word
+      current_node.isWord = true;
+    }
+
+    /* Insert a list of words. */
+    that.insertWords = function (words) {
+      for (var i = 0; i < words.length; i++) {
+        that.insertWord(words[i]);
+      }
+    }
+
+    /* Get the node that a word is located at. */
+    that.getNode = function (word) {
+      // Start at the root
+      var current_node = that;
+      for (var i = 0; i < word.length; i++) {
+        var c = word[i];
+
+        // If the word's character isn't a child of the current_node,
+        // The word isn't in the trie
+        if (!(c in current_node.children)) {
+          return;
+        }
+        // Move down the trie, update current_node
+        current_node = current_node.children[c];
+      };
+      return current_node;
+    }
+
+    /* Check if trie contains words. */
+    that.contains = function (word) {
+      var current_node = that.getNode(word);
+      if (current_node) {
+        return current_node.isWord;
+      }
+      return false;
+    }
+
+    /* Get a list of words that start with the same prefix, limit to count number. */
+    that.getWords = function (word, count) {
+      function fork(n, w) {
+        function child(c) {
+          return fork(n.children[c], w + c);
+        }
+
+        n.isWord && words.push(w);
+        return words.length >= count || Object.keys(n.children).some(child);
```

Comment:
what is this condition checking?

Reply:
For the first line: if the node is the last node in the word, then it pushes the word to the words array and is true.

For the second line: It checks to make sure that the number of words returned are less than the count specified. If this is true, then it gets the children. It then gets the keys from the node.children object and the some method will check if any of the children of the node satisfy the child(character) function. (Calls the child function on the node's children until one of them is the end of the word).

I added more comments to the function!